### PR TITLE
[DO NOT MERGE] chore: Make git status API faster

### PR DESCRIPTION
**Pull request by subrata71**
> Pull Request Template
>
> Use this template to quickly create a well written pull request. Delete all quotes before creating the pull request.
>
#### PR fixes following issue(s)
Fixes # (issue number)
> if no issue exists, please create an issue and ask the maintainers about this first
>
>
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Chore (housekeeping or task changes that don't impact user perception)
- This change requires a documentation update
>
>
>
- [ ] Manual
- [ ] JUnit
- [ ] Jest
- [ ] Cypress
>
>
#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

## Summary by CodeRabbit

- **New Features**
	- Introduced enhanced Git operation handling, including a new method for retrieving branch status with file locking support.
- **Improvements**
	- Updated Git status fetching logic to optimize the deployment and merge processes in the editor.
- **Refactor**
	- Refactored Git status retrieval process for better efficiency and error handling.

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/GitServiceCEImpl.java [hunk 1]
```diff
@@ -1953,25 +1944,40 @@ private Mono<GitStatusDTO> getStatus(
...
+                    return gitExecutor
+                            .getStatus(tuple.getT2().getT1(), finalBranchName)
+                            .elapsed()
+                            // Remove any files which are copied by hard resetting the repo
+                            .flatMap(tuple2 -> {
+                                log.debug("** git status took: {}", tuple2.getT1() - atomicLong.get());
+                                atomicLong.set(tuple2.getT1());
+                                try {
+                                    // TODO: Why are we doing reset to last commit always???
+                                    // TODO: We can get rid of that - Anagh
```
Comment (AnaghHegde) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/GitServiceCEImpl.java:1960]: I think we can remove this comment

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/GitServiceCEImpl.java [hunk 2]
```diff
@@ -1953,25 +1944,40 @@ private Mono<GitStatusDTO> getStatus(
...
+                                            branchName,
+                                            e);
+                                    return Mono.error(new AppsmithException(
+                                            AppsmithError.GIT_ACTION_FAILED, "status", e.getMessage()));
+                                }
+                            });
+                })
+                .elapsed()
+                .flatMap(tuple2 -> {
+                    log.debug("** reset to last commit took: {}", tuple2.getT1() - atomicLong.get());
```
Comment (AnaghHegde) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/GitServiceCEImpl.java:1978]: What is the significance of the ** in this log message?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/GitServiceCEImpl.java [hunk 3]
```diff
@@ -1994,11 +2000,184 @@ private Mono<GitStatusDTO> getStatus(
...
+                })
+                .tag("gitStatus", defaultApplicationId)
+                .name(AnalyticsEvents.GIT_STATUS.getEventName())
+                .tap(Micrometer.observation(observationRegistry));
+
+        return Mono.zip(statusMono, sessionUserService.getCurrentUser(), branchedAppMono)
+                .elapsed()
+                .flatMap(objects -> {
+                    Long elapsedTime = objects.getT1();
+                    log.debug(" ** Multi mono took: {}", elapsedTime);
```
Comment (AnaghHegde) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/GitServiceCEImpl.java:2180]: Same as above

**Review by AnaghHegde — CHANGES_REQUESTED** [review]
Since we are using the `.tag` to send this data to grafana do we really need to add this info again to logs? It will clutter the log files and makes it hard to read the other information while debugging.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/GitServiceCEImpl.java (hunk 2 above)
Comment (subrata71) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/GitServiceCEImpl.java:1978]: This was an experimental PR created during the hackathon. Closing this in favor of this [PR](https://github.com/appsmithorg/appsmith/pull/31229)

**Comment by subrata71** [comment]
Closing this in favor of this PR https://github.com/appsmithorg/appsmith/pull/31229

**Pull request closed without merging**